File: src/core/aggregator.js
'use strict';

// 聚合器：把三源的标准化用量事件合并、去重，按工具/模型/日汇总，并计价。
// 输出结构供 UI 与 CLI 直接消费。

const { costOfTokens } = require('../pricing/calculator');

// 新建一个空的用量桶。
function emptyBucket() {
  return {
    tokens: { input: 0, output: 0, cacheWrite: 0, cacheRead: 0 },
    total: 0, // 四类 token 之和
    cost: 0,
    estimated: false, // 是否含估算定价
  };
}

// 把事件的 token 累加进桶。
function addTokens(bucket, ev) {
  bucket.tokens.input += ev.input;
  bucket.tokens.output += ev.output;
  bucket.tokens.cacheWrite += ev.cacheWrite;
  bucket.tokens.cacheRead += ev.cacheRead;
  bucket.total += ev.input + ev.output + ev.cacheWrite + ev.cacheRead;
}

// 本地时区日期键 YYYY-MM-DD。
function localDayKey(ts) {
  const d = ts == null ? new Date() : new Date(ts);
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

// 计算本地时区“今日零点”的毫秒时间戳。
function startOfToday(now = Date.now()) {
  const d = new Date(now);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

// 生成以 end 为末日的连续 n 个本地日期键（含 end 当天，从早到晚）。
function listLastNDayKeys(n, endTs = Date.now()) {
  const end = new Date(endTs);
  end.setHours(0, 0, 0, 0);
  const keys = [];
  for (let i = n - 1; i >= 0; i--) {
    const d = new Date(end);
    d.setDate(end.getDate() - i);
    keys.push(localDayKey(d.getTime()));
  }
  return keys;
}

// 确保 byDay 中有该日桶。
function ensureDay(byDay, dayKey) {
  if (!byDay[dayKey]) {
    byDay[dayKey] = {
      ...emptyBucket(),
      tools: {},
    };
  }
  return byDay[dayKey];
}

// 确保日桶内有某工具的子桶。
function ensureDayTool(dayBucket, tool) {
  if (!dayBucket.tools[tool]) {
    dayBucket.tools[tool] = emptyBucket();
  }
  return dayBucket.tools[tool];
}

// 由 byDay 与当前时刻生成 period（今日 / 近 7 日 / 近 7 日序列）。
function buildPeriod(byDay, now = Date.now()) {
  const todayKey = localDayKey(now);
  const last7Keys = listLastNDayKeys(7, now);

  const sumKeys = (keys) => {
    const out = emptyBucket();
    for (const k of keys) {
      const b = byDay[k];
      if (!b) continue;
      out.tokens.input += b.tokens.input;
      out.tokens.output += b.tokens.output;
      out.tokens.cacheWrite += b.tokens.cacheWrite;
      out.tokens.cacheRead += b.tokens.cacheRead;
      out.total += b.total;
      out.cost += b.cost;
      if (b.estimated) out.estimated = true;
    }
    return { tokens: out.tokens, total: out.total, cost: out.cost, estimated: out.estimated };
  };

  const todayBucket = byDay[todayKey];
  const today = todayBucket
    ? {
        tokens: { ...todayBucket.tokens },
        total: todayBucket.total,
        cost: todayBucket.cost,
        estimated: todayBucket.estimated,
      }
    : { tokens: emptyBucket().tokens, total: 0, cost: 0, estimated: false };

  const days = last7Keys.map((date) => {
    const b = byDay[date];
    return {
      date,
      total: b ? b.total : 0,
      cost: b ? b.cost : 0,
    };
  });

  return {
    todayKey,
    today,
    last7: sumKeys(last7Keys),
    days,
  };
}

// 序列化 byDay，去掉内部可变引用。
function serializeByDay(byDay) {
  const out = {};
  for (const [date, b] of Object.entries(byDay)) {
    const tools = {};
    for (const [name, t] of Object.entries(b.tools || {})) {
      tools[name] = {
        tokens: { ...t.tokens },
        total: t.total,
        cost: t.cost,
        estimated: t.estimated,
      };
    }
    out[date] = {
      tokens: { ...b.tokens },
      total: b.total,
      cost: b.cost,
      estimated: b.estimated,
      tools,
    };
  }
  return out;
}

// 主聚合函数。events：已合并的全部事件；table：定价表。
// opts.now：可注入当前时间，便于测试“今日/近7日”边界。
function aggregate(events, table, opts = {}) {
  const seen = new Set();
  const now = opts.now != null ? opts.now : Date.now();
  const todayStart = startOfToday(now);
  const byDay = {};
  const tools = {};

  for (const ev of events) {
    // 统一去重（Claude 跨文件重复的关键防线）。
    if (ev.dedupeKey) {
      if (seen.has(ev.dedupeKey)) {
        continue;
      }
      seen.add(ev.dedupeKey);
    }

    if (!tools[ev.tool]) {
      tools[ev.tool] = {
        ...emptyBucket(),
        models: {},
        today: emptyBucket(),
        sessions: new Set(),
      };
    }
    const t = tools[ev.tool];
    const priced = costOfTokens(ev, ev.model, table);

    addTokens(t, ev);
    t.cost += priced.cost;
    if (priced.estimated) t.estimated = true;

    if (!t.models[ev.model]) {
      t.models[ev.model] = { ...emptyBucket(), matched: priced.matched, free: priced.free };
    }
    const m = t.models[ev.model];
    addTokens(m, ev);
    m.cost += priced.cost;
    if (priced.estimated) m.estimated = true;

    const ts = ev.ts != null ? ev.ts : now;
    if (ts >= todayStart) {
      addTokens(t.today, ev);
      t.today.cost += priced.cost;
      if (priced.estimated) t.today.estimated = true;
    }

    // 按本地日累计（跨日历史曲线的数据源）。
    const dayKey = localDayKey(ts);
    const dayBucket = ensureDay(byDay, dayKey);
    addTokens(dayBucket, ev);
    dayBucket.cost += priced.cost;
    if (priced.estimated) dayBucket.estimated = true;
    const dayTool = ensureDayTool(dayBucket, ev.tool);
    addTokens(dayTool, ev);
    dayTool.cost += priced.cost;
    if (priced.estimated) dayTool.estimated = true;

    if (ev.sessionId) {
      t.sessions.add(ev.sessionId);
    }
  }

  // 收尾：sessions 转计数，拼总览与 period。
  const grand = { ...emptyBucket() };
  const toolsOut = {};
  for (const [name, t] of Object.entries(tools)) {
    toolsOut[name] = {
      tokens: t.tokens,
      total: t.total,
      cost: t.cost,
      estimated: t.estimated,
      today: {
        tokens: t.today.tokens,
        total: t.today.total,
        cost: t.today.cost,
        estimated: t.today.estimated,
      },
      sessionCount: t.sessions.size,
      models: t.models,
    };
    grand.tokens.input += t.tokens.input;
    grand.tokens.output += t.tokens.output;
    grand.tokens.cacheWrite += t.tokens.cacheWrite;
    grand.tokens.cacheRead += t.tokens.cacheRead;
    grand.total += t.total;
    grand.cost += t.cost;
    if (t.estimated) grand.estimated = true;
  }

  const byDayOut = serializeByDay(byDay);
  const period = buildPeriod(byDayOut, now);

  return {
    generatedAt: now,
    tools: toolsOut,
    grand,
    byDay: byDayOut,
    period,
  };
}

module.exports = {
  aggregate,
  emptyBucket,
  startOfToday,
  localDayKey,
  listLastNDayKeys,
  buildPeriod,
};

File: src-tauri/src/core/aggregator.rs
use super::types::*;
use super::pricing;
use std::collections::{HashMap, HashSet};

/// 聚合事件列表为 Snapshot
pub fn aggregate(events: &[TokenEvent]) -> Snapshot {
    let generated_at = chrono::Utc::now().timestamp_millis();
    let today_key = chrono::Local::now().format("%Y-%m-%d").to_string();

    let mut by_day: HashMap<String, DayData> = HashMap::new();
    let mut tools: HashMap<String, ToolStats> = HashMap::new();
    let mut grand = GrandTotal::default();
    let mut seen_dedupe = HashSet::new();
    let mut sessions: HashMap<String, HashSet<String>> = HashMap::new();

    for event in events {
        // 去重
        if !seen_dedupe.insert(event.dedupe_key.clone()) {
            continue;
        }

        // 按工具分组
        let tool_entry = tools.entry(event.tool.clone()).or_insert_with(|| ToolStats {
            tokens: ToolTokens::default(),
            total: 0,
            cost: 0.0,
            estimated: false,
            session_count: 0,
            models: HashMap::new(),
            today: TodayStats::default(),
        });

        add_to_tokens(&mut tool_entry.tokens, event);
        tool_entry.total = tool_entry.tokens.total();
        tool_entry.cost += event.cost;
        if event.estimated {
            tool_entry.estimated = true;
        }

        // 今日统计
        if event.day_key == today_key {
            add_to_tokens(&mut tool_entry.today.tokens, event);
            tool_entry.today.total = tool_entry.today.tokens.total();
            tool_entry.today.cost += event.cost;
            if event.estimated {
                tool_entry.today.estimated = true;
            }
        }

        // 按模型分组
        let model_entry = tool_entry
            .models
            .entry(event.model.clone())
            .or_insert_with(|| {
                let matched = pricing::match_model(&event.model);
                let is_est = !pricing::is_free(&matched);
                ModelStats {
                    tokens: ToolTokens::default(),
                    total: 0,
                    cost: 0.0,
                    estimated: is_est,
                    matched: matched.clone(),
                    free: pricing::is_free(&matched),
                }
            });

        add_to_tokens(&mut model_entry.tokens, event);
        model_entry.total = model_entry.tokens.total();
        model_entry.cost += event.cost;

        // 会话去重
        if !event.session_id.is_empty() {
            sessions
                .entry(event.tool.clone())
                .or_default()
                .insert(event.session_id.clone());
        }

        // 按天分组
        let day_entry = by_day.entry(event.day_key.clone()).or_insert_with(|| DayData {
            tokens: ToolTokens::default(),
            total: 0,
            cost: 0.0,
            estimated: false,
            tools: HashMap::new(),
        });

        add_to_tokens(&mut day_entry.tokens, event);
        day_entry.total = day_entry.tokens.total();
        day_entry.cost += event.cost;
        if event.estimated {
            day_entry.estimated = true;
        }

        // 按天的工具子分组
        let day_tool = day_entry
            .tools
            .entry(event.tool.clone())
            .or_insert_with(ToolTokens::default);
        add_to_tokens(day_tool, event);

        // Grand total
        add_to_tokens(&mut grand.tokens, event);
        grand.total = grand.tokens.total();
        grand.cost += event.cost;
        if event.estimated {
            grand.estimated = true;
        }
    }

    // 填充 session_count
    for (tool, session_set) in &sessions {
        if let Some(ts) = tools.get_mut(tool) {
            ts.session_count = session_set.len() as u64;
        }
    }

    // 计算 period
    let period = build_period(&by_day, &today_key);

    // 来源状态
    let sources = SourcesResult::default();

    Snapshot {
        generated_at,
        tools,
        grand,
        by_day,
        period,
        sources,
    }
}

fn add_to_tokens(tokens: &mut ToolTokens, event: &TokenEvent) {
    tokens.input += event.input;
    tokens.output += event.output;
    tokens.cache_write += event.cache_write;
    tokens.cache_read += event.cache_read;
    tokens.total = tokens.total();
}

impl ToolTokens {
    fn total(&self) -> u64 {
        self.input + self.output + self.cache_write + self.cache_read
    }
}

/// 构建周期统计
fn build_period(
    by_day: &HashMap<String, DayData>,
    today_key: &str,
) -> PeriodData {
    let mut last7_tokens = ToolTokens::default();
    let mut last7_cost = 0.0;
    // 生成最近 7 天的键
    let today = chrono::Local::now().date_naive();
    let mut days = Vec::new();

    for i in (0..7).rev() {
        let d = today - chrono::Duration::days(i);
        let key = d.format("%Y-%m-%d").to_string();
        if let Some(day_data) = by_day.get(&key) {
            add_tokens_to_tool(&mut last7_tokens, day_data);
            last7_cost += day_data.cost;
            days.push(DaySummary {
                date: key.clone(),
                total: day_data.total,
                cost: day_data.cost,
            });
        } else {
            days.push(DaySummary {
                date: key,
                total: 0,
                cost: 0.0,
            });
        }
    }

    // 今日
    let default_day = DayData::default();
    let today_data = by_day.get(today_key).unwrap_or(&default_day);
    let today_tokens = ToolTokens {
        input: today_data.tokens.input,
        output: today_data.tokens.output,
        cache_write: today_data.tokens.cache_write,
        cache_read: today_data.tokens.cache_read,
        total: today_data.total,
    };

    PeriodData {
        today_key: today_key.to_string(),
        today: today_tokens,
        today_cost: today_data.cost,
        last7: last7_tokens,
        last7_cost,
        days,
    }
}

fn add_tokens_to_tool(target: &mut ToolTokens, source: &DayData) {
    target.input += source.tokens.input;
    target.output += source.tokens.output;
    target.cache_write += source.tokens.cache_write;
    target.cache_read += source.tokens.cache_read;
    target.total = target.total();
}
